authz/src/authorizer.rs

```diff
@@ -0,0 +1,57 @@
+use async_trait::async_trait;
+
+use super::{Error, Permission};
+
+/// An authorizer is used to validate the associated with
+/// an authorization token that has been extracted from a request.
+#[async_trait]
+pub trait Authorizer: std::fmt::Debug + Send + Sync {
+    /// Determine the permissions associated with a request token.
+    ///
+    /// The returned list of permissions is the intersection of the permissions
+    /// requested and the permissions associated with the token.
+    ///
+    /// Implementations of this trait should only error if:
+    ///     * there is a failure processing the token.
+    ///     * there is not any intersection of permissions.
+    async fn permissions(
+        &self,
+        token: Option<Vec<u8>>,
+        perms: &[Permission],
+    ) -> Result<Vec<Permission>, Error>;
+
+    /// Make a test request that determines if end-to-end communication
+    /// with the service is working.
+    async fn probe(&self) -> Result<(), Error> {
+        self.permissions(Some(b"".to_vec()), &[]).await?;
+        Ok(())
```

Comment:
I think this needs to become something along the lines of:

```suggestion
        match self.permissions(Some(b"".to_vec()), &[]).await {
            Ok(_) | Err(Error::Forbidden) | Err(Error::InvalidToken) => Ok(()),
            Err(e) => Err(e),
        }
```